### PR TITLE
WIP: GIVC TLS

### PR DESCRIPTION
**Pull request by mbssrc**
## Description of changes

- change networking into two different definitions to prepare network removal from host
- improve `config.networking` integration
- add and populate givc tls module
- add givc-cli wrapper module for easier use in UI
- remove logging from x86 profile

## Checklist for things done

- [ ] Summary of the proposed changes in the PR description
- [ ] More detailed description in the commit message(s)
- [ ] Commits are squashed into relevant entities - avoid a lot of minimal dev time commits in the PR
- [ ] [Contribution guidelines](https://github.com/tiiuae/ghaf/blob/main/CONTRIBUTING.md) followed
- [ ] Ghaf documentation updated with the commit - https://tiiuae.github.io/ghaf/
- [ ] PR linked to architecture documentation and requirement(s) (ticket id)
- [ ] Test procedure described (or includes tests). Select one or more:
  - [ ] Tested on Lenovo X1 `x86_64`
  - [ ] Tested on Jetson Orin NX or AGX `aarch64`
  - [ ] Tested on Polarfire `riscv64`
- [ ] Author has run `make-checks` and it passes
- [ ] All automatic Github Action checks pass - see [actions](https://github.com/tiiuae/ghaf/actions)
- [ ] Author has added reviewers and removed PR draft status
- [ ] Change requires full re-installation
- [ ] Change can be updated with `nixos-rebuild ... switch`

## Instructions for Testing

- [ ] List all targets that this applies to:
- [ ] Is this a new feature
  - [ ] List the test steps to verify:
- [ ] If it is an improvement how does it impact existing functionality?

### CODE REVIEW

File: modules/common/services/desktop.nix [hunk 1]
```diff
@@ -40,17 +40,6 @@ in
       profiles.graphics.compositor = "labwc";
       graphics = {
         launchers =
-          let
```
Comment (dmitry-erin) [inline modules/common/services/desktop.nix:43]: They might be needed for ctrl-panel as well, at least ´--notls´ option

Comment (mbssrc) [reply modules/common/services/desktop.nix:43]: Yes, I can add a wrapper for ctrl-panel as well

File: modules/microvm/virtualization/microvm/modules.nix [hunk 2]
```diff
@@ -46,11 +46,17 @@ let
     inherit (config.ghaf.qemu) audiovm;
   };
 
+  # Common namespace to pass parameters at built-time from host to VMs
+  commonModule = {
+    config.ghaf.common = config.ghaf.common;
```
Comment (brianmcgillion) [inline modules/microvm/virtualization/microvm/modules.nix:51]: inherit?

File: modules/givc/common.nix [hunk 3]
```diff
@@ -1,18 +1,50 @@
...
+        description = "Name of the IP range for parsing";
+        type = types.str;
+      };
+      addr = mkOption {
+        description = "IP address of admin server";
+        type = types.str;
+      };
+      port = mkOption {
+        description = "Port of admin server";
+        type = types.str;
```
[the remark this answers is not in the document]
Comment (mbssrc) [reply modules/givc/common.nix:29]: May change this later in sync with givc

**Comment by mbssrc** [comment]
[39b4699](https://github.com/tiiuae/ghaf/pull/915/commits/39b46993c14ef15fae6be90c086b4b191d927ed6) Updates:
- rebase to latest
- address review comments
- change the way keys/certificates are packaged. Now they are stored in `/storagevm/givc` folder as images, so they can be mounted read-only into VMs. This allows to preserve "stateless" VMs such as zathura.
- move the host folder generation to no be guarded by unsafe shares

**Review by brianmcgillion — APPROVED** [review]

**Comment by johannarautanen** [comment]
**Checked with native Orin AGX:**

Issue:
- cannot sign in with Element although is internet connection available
Screenshot from 2025-01-30 07-37-16

Working:
- can only connect with net-vm or ghaf-host. net-vm-debug or ghaf-host-debug are not working anymore
- IP's 192.168.100.1 and 192.168.100.2 working. OId IP's not working anymore (192.168.101.1 or 192.168.101.2)
- regression tests working
- all 4 application can be launched

**Comment by johannarautanen** [comment]
**Checked with native Orin NX:**

Issues:
- there is weird thing ongoing with time sync. First time was ok, then i switch down the NX for awhile and switch on again. Time wasn't via internet connection to time, time was about 40 minutes behind the right time.
- cannot sign in with Element although is internet connection available

Working:
- can only connect with net-vm or ghaf-host. net-vm-debug or ghaf-host-debug are not working anymore
- IP's 192.168.100.1 and 192.168.100.2 working. OId IP's not working anymore (192.168.101.1 or 192.168.101.2)
- all 4 application can be launched

**Comment by milva-unikie** [comment]
**Tested on Lenovo-X1** (`full re-installation`)

**Issues:**
- Grafana logs are only getting sent from admin-vm, no logs from other vm:s.
- Laptop gets sometimes stuck at reboot by power menu. I can see the nixos bootloader and some bootlogs, but after that the screen goes black and login screen never appears. Cannot connect to host by ssh after this happened.

**Working:**

- All vm:s are available, debug names don't work anymore.
- GIVC certificates are available.
- Apps seem to be working.
- Audio works, cameras work.
- VPN and other connections inside business-vm work.
- Shared folders and file handlers work.
- Bumps control panel from 15.11.2024 -> 12.12.2024. I only checked that the old features (Timezone change & Display scale) work. There is point to check & report issues that might have been already fixed a month ago.

**Notes:**

- There are fails in pre-merge-pipeline for X1 because the net-vm ip address was changed. PR to fix that  https://github.com/tiiuae/ci-test-automation/pull/222. These need to be merged at the same time.
- Host system clock is not getting synchronized with ethernet, only with wifi. Still unclear what the desired way is so waiting for confirmation on that. If it stays as it is now we will need to update tests to adapt to the change.

**Comment by mbssrc** [comment]
## Update [2f736d9](https://github.com/tiiuae/ghaf/pull/915/commits/2f736d96528720766d0848d6bab1875db6db39e1)

### General
USB ethernet dongles (which are used for testing) are now passed to the net-vm on all devices.

Usage of `nixos-rebuild` requires a proxy jump. This can be done by

- adding a proxy-jump ssh configuration to your dev machine, such as
```
programs = {
  ssh = {
    startAgent = true;
    extraConfig = ''
      host ghaf-netvm
        user root
        hostname <target-ip>
      host ghaf-host
         user root
         hostname 192.168.100.2
         proxyjump ghaf-netvm
    '';
  };
};
```

- export the NIX_SSHOPTS environment variable before using nixos-rebuild:
`export NIX_SSHOPTS="-o ProxyJump=root@<your-target-ip>"`

- using the new helper in the Ghaf devshell `ghaf-rebuild` that does it for you. 
Usage: `ghaf-rebuild <taget machine ip> <build target> [regular nixos-rebuild options]`
Example: `ghaf-rebuild 192.168.0.123 .#lenovo-x1-carbon-gen11-debug switch`

This applies to all targets. 

**Exceptions:**
Connecting to the host directly is possible in the following cases:
- AGX if connected via physical ethernet port
- X1 and AGX/NX if net-vm is stopped (ethernet dongle should then be attached to the host)


### AGX/NX notes

This should resolve the errors encountered with the network connection. 
On AGX, if connected via the physical ethernet port only, timesync and element will fail as no there is no ethernet passthrough to the net-vm, but the route is still in place.

### Lenovo x1 notes

- Logging address fixed, should behave correctly now
- Host clock should be synchronized now as to above fix

**Comment by mbssrc** [comment]
Rebased on latest main [d7bd905](https://github.com/tiiuae/ghaf/pull/915/commits/d7bd9054d1b841ac1f1a57e5b854e078ee158c10)

**Comment by mbssrc** [comment]
[beb1454](https://github.com/tiiuae/ghaf/pull/915/commits/beb14549d9e745e0172bb09bb279698675a444b1)
Fix packet forwarder ip/mac addressing after rebase

**Comment by mbssrc** [comment]
[4a36ab4](https://github.com/tiiuae/ghaf/pull/915/commits/4a36ab49f0897c4049ca18760e84547e155577b3)
another rebase

File: packages/ghaf-build-helper/default.nix [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+      proxy_jump="$1"
+    else
+      exit 1
+    fi
+
+    build_target="$2"
+    shift 2
+
+    export NIX_SSHOPTS="-o ProxyJump=root@$proxy_jump"
+    nixos-rebuild --flake "$build_target" --target-host root@ghaf-host --fast "$@"
```
Comment (brianmcgillion) [inline packages/ghaf-build-helper/default.nix:69]: ghaf-host? deus ex machina?

**Review by brianmcgillion — DISMISSED** [review]

**Comment by johannarautanen** [comment]
**Checked with native Orin AGX:**

**Issue:**
- cannot sign in with Element although is internet connection(ethernet) available. Can't attached the journalctl-logs, cause cannot connect to the AGX from remote laptop.

**Comment by johannarautanen** [comment]
**Checked with native Orin NX:**

- When attached the ethernet to the physical port (net-vm) the Ghaf time updated.
- can sign to the element-app
- and when the net-vm is closed via ghaf-host, you can sign  to the Element when USB-ethernet adapter (ghaf-host) is used.

**Comment by milva-unikie** [comment]
**Tested on Lenovo-X1** (`full re-installation`)

New issues:
- VPN does not work any more. I am able to connect to VPN with the app but can't access jira.tii.ae in Trusted Browser.

Previous issues have been fixed:
- Host clock is now synchronized.
- Logs are sent to Grafana.

Notes:
- I confirmed that the boot issue I reported earlier is not related to this pr.
- Test-automation fix has been updated to the latest changes. Most of the bat-tests pass on X1, but many still need more reworking. Orins have not yet been tested. https://github.com/tiiuae/ci-test-automation/pull/222

**Comment by mbssrc** [comment]
> **Tested on Lenovo-X1** (`full re-installation`)
> 
> New issues:
> 
>     * VPN does not work any more. I am able to connect to VPN with the app but can't access jira.tii.ae in Trusted Browser.
> 
> 
> Previous issues have been fixed:
> 
>     * Host clock is now synchronized.
> 
>     * Logs are sent to Grafana.
> 
> 
> Notes:
> 
>     * I confirmed that the boot issue I reported earlier is not related to this pr.
> 
>     * Test-automation fix has been updated to the latest changes. Most of the bat-tests pass on X1, but many still need more reworking. Orins have not yet been tested. [Update network connections ci-test-automation#222](https://github.com/tiiuae/ci-test-automation/pull/222)

Cannot reproduce the VPN error on my side. Also can't think of a reason how it could be affected. Please check again.

File: modules/common/common.nix [hunk 5]
```diff
@@ -14,7 +28,59 @@
...
+        type = types.listOf types.str;
+        default = [ ];
+        description = "List of VMs currently enabled.";
+      };
+      systemHosts = mkOption {
+        type = types.listOf types.str;
+        default = [ ];
+        description = "List of system hosts currently enabled.";
+      };
+      appHosts = mkOption {
```
[the remark this answers is not in the document]
Comment (mbssrc) [reply modules/common/common.nix:43]: Since their purpose is to be used for auto-generation, not sure what assertions would do here. Added  `readOnly`

**Comment by milva-unikie** [comment]
> Cannot reproduce the VPN error on my side. Also can't think of a reason how it could be affected. Please check again.

tldr: works now

It definitely was not working yesterday or this morning. I re-installed and still nothing. And then suddenly it started working and has now worked for an hour. After it started working I tried again to re-install and this time it worked right away. No idea what happened but should be good now. Sorry about the confusion.

File: modules/common/common.nix (hunk 5 above)
[the remark this answers is not in the document]
Comment (mbssrc) [reply modules/common/common.nix:43]: reverted it as it does not build on nvidia

**Review by brianmcgillion — APPROVED** [review]
